source/http2_stream_manager.c

```diff
@@ -636,6 +776,22 @@ static void s_on_stream_complete(struct aws_http_stream *stream, int error_code,
         pending_stream_acquisition->options.on_complete(
             stream, error_code, pending_stream_acquisition->options.user_data);
     }
+    if (stream_manager->close_connection_on_server_error && error_code == AWS_ERROR_SUCCESS) {
+        /* Check status code if stream completed successfully. */
+        int status_code = 0;
+        aws_http_stream_get_incoming_response_status(stream, &status_code);
+        AWS_ASSERT(status_code != 0); /* The get status should not fail */
+        if (status_code / 100 == 5) {
+            STREAM_MANAGER_LOGF(
+                DEBUG,
+                stream_manager,
+                "%d response received for stream: %p. Closing connection: %p",
+                status_code,
+                (void *)stream,
+                (void *)sm_connection->connection);
+            aws_http_connection_close(sm_connection->connection);
```

Comment:
I see a few problems with this approach:
1) major: There might be other streams on this connection, and they will get suddenly murdered due to the connection shutting down.
2) minor: We could have stopped putting new streams on this connection earlier (when the headers were first received) instead of waiting until the response was complete

Can we use `aws_http2_send_goaway()` instead? I think (???) that would allow us to finish any existing streams, while preventing any new streams from getting put on this connection. Also, we might be able to call aws_http2_send_goaway() earlier, from the on_headers callback, instead of waiting until on_complete

Reply:
We don't prevent new streams if we send goaway. The spec also doesn't mentioned the sender behavior after sending goaway beside not receiving any new stream from other side.

I am not sure about should we enforce people to be like after sending goaway, no more new streams can be made... It seems making sense though... Will do that...

Reply:
If you want the ping stuff to be merged before the server-error stuff, you might want to split this into 2 different PRs 🤷‍♀️

Reply:
added `stop_new_request` api